src/eve/typingx.py

```diff
@@ -218,3 +222,134 @@ def resolve_type(
                 raise e
 
     return actual_type
+
+
+def _collapsable_type_args(*args: Any) -> tuple[bool, tuple]:
+    if len(args) and all(args[0] == a for a in args):
+        return (True, args)
+    else:
+        return (False, args)
+
+
+@dataclasses.dataclass
+class CallableKwargsInfo:
+    data: dict[str, Any]
+
+
+@functools.singledispatch
+def get_typing(value: Any, *, annotate_callable_kwargs: bool = False) -> Any:
+    """Generate a typing definition from a value.
+
+    The implementation uses :func:`functools.singledispatch`. Customized or
+    more detailed annotations can be generated by registering alternative
+    implementations.
+
+    Examples:
+        >>> get_typing(3)
+        int
+
+        >>> get_typing((3, "four"))
+        tuple[int, str]
+
+        >>> get_typing((3, 4))
+        tuple[int, ...]
+
+        >>> get_typing(frozenset([1, 2, 3]))
+        frozenset[int]
+
+        >>> get_typing({'a': 0, 'b': 1})
+        dict[str, int]
+
+        >>> get_typing({'a': 0, 'b': 'B'})
+        dict[str, typing.Any]
+
+        >>> get_typing(lambda a, b: a + b)
+        collections.abc.Callable[[typing.Any, typing.Any], typing.Any]
+
+        >>> def f(a: int, b) -> int: ...
+        >>> get_typing(f)
+        collections.abc.Callable[['int', typing.Any], 'int']
+
+        >>> def f(a: int, b) -> int: ...
+        >>> get_typing(f)
+        collections.abc.Callable[..., 'int']
+
+        >>> get_typing(Dict[int, Union[int, float]])
+        dict[int, typing.Union[int, float]]
+
+        >>> get_typing(Dict[int, Union[TypeVar("T", bound=int), float]])
+        dict[int, typing.Union[int, float]]
+
+        >>> get_typing(Dict[int, Union[int, float]], get_origins=False)
+        typing.Dict[int, typing.Union[int, float]]
+
+        >>> import numbers
+        >>> @get_typing.register(int)
+        >>> @get_typing.register(float)
+        >>> @get_typing.register(complex)
+        >>> def _get_typing_number(value, *, get_origins: bool = True):
+        ...    return numbers.Number
+        >>> get_typing(3.4)
+        numbers.Number
+
+    """
+    recursive_get = functools.partial(get_typing, annotate_callable_kwargs=annotate_callable_kwargs)
+    if isinstance(value, type):
+        return value
+
+    elif value in (None, types.NoneType):
+        return None
+
+    elif isinstance(value, tuple):
+        unique_type, args = _collapsable_type_args(*(recursive_get(item) for item in value))
+        if unique_type and len(args) > 1:
+            return tuple[args[0], ...]
+        elif len(args):
+            return tuple[args]
```

Comment:
My intuition would have been that it should be `tuple[*args]`

Reply:
That would be a syntax error since you cannot unpack inside a `[]` operator expression. I guess it's because `getitem` only accepts a single argument (although you can use unpacking if you call `__getitem__` directly, but that would cause a run-time error)